keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt

```diff
@@ -0,0 +1,144 @@
+
+import com.fasterxml.jackson.module.kotlin.readValue
+import com.netflix.spinnaker.keel.api.ApiVersion
+import com.netflix.spinnaker.keel.api.ArtifactType
+import com.netflix.spinnaker.keel.api.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.name
+import com.netflix.spinnaker.keel.api.randomUID
+import com.netflix.spinnaker.keel.api.uid
+import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.keel.persistence.NoSuchDeliveryConfigName
+import com.netflix.spinnaker.keel.persistence.metamodel.Tables.DELIVERY_ARTIFACT
+import com.netflix.spinnaker.keel.persistence.metamodel.Tables.DELIVERY_CONFIG
+import com.netflix.spinnaker.keel.persistence.metamodel.Tables.DELIVERY_CONFIG_ARTIFACT
+import com.netflix.spinnaker.keel.persistence.metamodel.Tables.ENVIRONMENT
+import com.netflix.spinnaker.keel.persistence.metamodel.Tables.ENVIRONMENT_RESOURCE
+import com.netflix.spinnaker.keel.persistence.metamodel.Tables.RESOURCE
+import com.netflix.spinnaker.keel.serialization.configuredObjectMapper
+import org.jooq.DSLContext
+
+class SqlDeliveryConfigRepository(
+  private val jooq: DSLContext,
+  private val resourceTypeIdentifier: (String) -> Class<*>
+) : DeliveryConfigRepository {
+
+  private val mapper = configuredObjectMapper()
+
+  override fun store(deliveryConfig: DeliveryConfig) {
+    jooq.inTransaction {
+      val uid = randomUID().toString()
+      with(deliveryConfig) {
+        insertInto(DELIVERY_CONFIG)
+          .set(DELIVERY_CONFIG.UID, uid)
+          .set(DELIVERY_CONFIG.NAME, name)
+          .set(DELIVERY_CONFIG.APPLICATION, application)
+          .onDuplicateKeyIgnore()
+          .execute()
+        artifacts.forEach { artifact ->
+          insertInto(DELIVERY_ARTIFACT)
+            .set(DELIVERY_ARTIFACT.UID, randomUID().toString())
+            .set(DELIVERY_ARTIFACT.NAME, artifact.name)
+            .set(DELIVERY_ARTIFACT.TYPE, artifact.type.name)
+            .onDuplicateKeyIgnore()
+            .execute()
+          insertInto(DELIVERY_CONFIG_ARTIFACT)
+            .set(DELIVERY_CONFIG_ARTIFACT.DELIVERY_CONFIG_UID, uid)
+            .set(DELIVERY_CONFIG_ARTIFACT.ARTIFACT_UID, select(DELIVERY_ARTIFACT.UID).from(DELIVERY_ARTIFACT).where(DELIVERY_ARTIFACT.NAME.eq(artifact.name)))
+            .onDuplicateKeyIgnore()
+            .execute()
+        }
+        environments.forEach { environment ->
+          val environmentUID = randomUID().toString()
+          insertInto(ENVIRONMENT)
+            .set(ENVIRONMENT.UID, environmentUID)
+            .set(ENVIRONMENT.DELIVERY_CONFIG_UID, uid)
+            .set(ENVIRONMENT.NAME, environment.name)
+            .onDuplicateKeyIgnore()
+            .execute()
+          environment.resources.forEach { resource ->
+            insertInto(RESOURCE)
```

Comment:
The fact that I'm having to reference tables "managed" by other repository classes here is why I wanted to start using the JOOQ code generator. I'd like to make some of this more modular as I take it forward but I didn't want to be repeating schema definitions across repository classes or just using raw string table / column names. Creating a standalone set of constants felt like I was re-inventing JOOQ functionality.